README.md
# ambari-chef

[![Build Status](https://travis-ci.org/hansohn/ambari-chef.svg?branch=master)](https://travis-ci.org/hansohn/ambari-chef) [![GitHub tag](https://img.shields.io/github/tag/hansohn/ambari-chef.svg)](https://github.com/hansohn/ambari-chef)

This cookbook installs [Apache Ambari](https://ambari.apache.org/), a project aimed at making Hadoop management simpler by developing software for provisioning, managing, and monitoring Apache Hadoop clusters. Ambari provides an intuitive, easy-to-use Hadoop management web UI backed by its RESTful APIs.

### Prerequsites

Ambari requires [Java](http://www.oracle.com/technetwork/java/javase/downloads/index.html) and [Python](https://www.python.org/) as prerequisites for installation. These prerequisites can be fulfilled using the `ambari-chef::ambari_server_prerequisites` or `ambari-chef::ambari_agent_prerequisites` recipes include with this cookbook or by any other means independently. You can customize the included prerequisites by modifying the following attributes.

```ruby
# python 2
override['python'] = {
  'python2' => {
    'packages' => ['python'],
  },
}

# java 8
override['java'] = {
  'install_from' => 'amazon_source',
  'install_version' => 'jdk-8u232-linux-x64',
}
```

### Configuration

By default this cookbook installs Ambari version ```2.7.4```, which at the time of this writing, is the current version. A different version of Ambari can be specified for installation by overriding the version attribute.

```ruby
# ambari
override['hw'] = {
  'ambari' => {
    'version' => '2.7.4',
    'server' => {
      'config' => {
        'ambari.properties' => {
          'api.ssl' => 'false',
          'client.api.port' => '8080',
        },
      },
    },
  },
}
```

### LDAP Authentication

To enable LDAP Authentication, define the following keys in your attributes file.

```ruby
# ambari ldap integration
override['hw'] = {
  'ambari' => {
    'server' => {
      'config' => {
        'ambari.properties' => {
          'ambari.ldap.isConfigured' => 'false',
          'authentication.ldap.baseDn' = 'dc=prd,dc=domain,dc=local',
          'authentication.ldap.managerDn' => 'uid=manager,cn=users,cn=accounts,dc=prd,dc=domain,dc=local',
          'authentication.ldap.primaryUrl' => 'ldap-server-01.prd.domain.local:636',
          'authentication.ldap.bindAnonymously' => 'false',
          'authentication.ldap.dnAttribute' => 'dn',
          'authentication.ldap.groupMembershipAttr' => 'memberUid',
          'authentication.ldap.groupNamingAttr' => 'cn',
          'authentication.ldap.groupObjectClass' => 'posixgroup',
          'authentication.ldap.managerPassword' => '/etc/ambari-server/conf/ldap-password.dat',
          'authentication.ldap.referral' => 'follow',
          'authentication.ldap.usernameAttribute' => 'uid',
          'authentication.ldap.userObjectClass' => 'person',
          'authentication.ldap.useSSL' => 'true',
        },
      },
    },
  },
}

# ambari ldap allowed users
override['hw'] = {
  'ambari' => {
    'server' => {
      'setup' => {
        'ldap' => {
          'ldap_sync_groups' => ['prd_user_group'],
          'ldap_sync_users' => ['prd_user'],
        },
      },
    },
  },
}
```

### PostgreSQL Database

Ambari installs and utilizes PostgreSQL by default. The `postgresql-server` package included in CestOS 7 at the time of this writing is `9.2.23`. If you are currently using a newer version of PostgreSQL you can modify the following attributes so that Ambari will point to your alternate version. You can can find more information about this [here](https://docs.hortonworks.com/HDPDocuments/Ambari-2.6.2.2/bk_ambari-administration/content/using_ambari_with_postgresql.html).

```ruby
# ambari server db setup values
override['hw'] = {
  'ambari' => {
    'server' => {
      'setup' => {
        'db' => {
          'databasehost' => 'localhost',
          'databaseport' => '5432',
          'databasepassword' => 'bigdata',
        },
      },
    },
  },
}

# ambari server config values
override['hw'] = {
  'ambari' => {
    'server' => {
      'config' => {
        'ambari.properties' => {
          'server.jdbc.connection-pool' => 'internal',
          'server.jdbc.database_name' => 'ambari',
          'server.jdbc.database' => 'postgres',
          'server.jdbc.driver' => 'org.postgresql.Driver',
          'server.jdbc.hostname' => 'localhost',
          'server.jdbc.port' => 5432,
          'server.jdbc.postgres.schema' => 'ambari',
          'server.jdbc.rca.driver' => 'org.postgresql.Driver',
          'server.jdbc.rca.url' => 'jdbc:postgresql://localhost:5432/ambari',
          'server.jdbc.rca.user.name' => 'ambari',
          'server.jdbc.rca.user.passwd' => '/etc/ambari-server/conf/password.dat',
          'server.jdbc.url' => 'jdbc:postgresql://localhost:5432/ambari',
          'server.jdbc.user.name' => 'ambari',
          'server.jdbc.user.passwd' => '/etc/ambari-server/conf/password.dat',
        },
      },
    },
  },
}
```

### Usage

##### Ambari Server

To install Ambari Server, include the ambari_server recipe in your run list

```ruby
# include ambari_server
include 'ambari-chef::ambari_server'
```

##### Ambari Agent

To install Ambari Server, include the ambari_agent recipe in your run list

```ruby
# include ambari_agent
include 'ambari-chef::ambari_agent'
```

##### Management Portal

Once installed, Ambari is available at http://127.0.0.1:8080 unless otherwise modified using the attributes referenced above. Ambari's default username and password are both `admin`.

### Documentation

The following resources may be helpful to better understand Ambari:

- [Apache Ambari](https://ambari.apache.org/)
- [Apache Ambari Blueprints](https://cwiki.apache.org/confluence/display/AMBARI/Blueprints)
- [Ambari Support Matrix](https://supportmatrix.hortonworks.com/)
- [Hortonworks Documentation](https://docs.hortonworks.com/index.html)
- [Hortonworks Community](https://community.hortonworks.com/answers/index.html)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](unreleased)

- no new features in development at this time

## [1.1.1](https://github.com/hansohn/ambari-chef/compare/1.1.0...1.1.1) (Dec 18, 2019)

BUG FIXES:

- update ambari agent sudoers file

## [1.1.0](https://github.com/hansohn/ambari-chef/compare/1.0.2...1.1.0) (Dec 17, 2019)

FEATURES:

- update default ambari version to 2.7.4
- update default java version to jdk-8u232-linux-x64 

## [1.0.2](https://github.com/hansohn/ambari-chef/compare/1.0.0...1.0.2) (May 31, 2019)

BUG FIXES:

- update attrbitute types to fix recipe inclusion and attribute overrides

## [1.0.0](https://github.com/hansohn/ambari-chef/compare/0.5.0...1.0.0) (May 29, 2019)

FEATURES:

- update attrbiute structure

BUG FIXES:

- call ambari version attributes at runtime

## [0.5.0](https://github.com/hansohn/ambari-chef/compare/0.4.1...0.5.0) (May 26, 2019)

FEATURES:

- just amazon corretto for java
- bump amabari version to 2.7.3

## [0.4.1](https://github.com/hansohn/ambari-chef/compare/0.4.0...0.4.1) (Aug 20, 2018)

BUG FIXES:

- update ssl password file name from http.pass.txt to https.pass.txt
- remove http.pass.txt file from ldap integration

## [0.4.0](https://github.com/hansohn/ambari-chef/compare/0.3.2...0.4.0) (Jul 24, 2018)

FEATURES:

- support version 2.4.0

## [0.3.2](https://github.com/hansohn/ambari-chef/compare/0.3.1...0.3.2) (Jul 23, 2018)

BUG FIXES:

- resolve issue where ambari version attribute is not intrepreted correctly when defined externally.

## [0.3.1](https://github.com/hansohn/ambari-chef/compare/0.3.0...0.3.1) (Jul 19, 2018)

BUG FIXES:

- grant missing ambari-agent sudo commands

## [0.3.0](https://github.com/hansohn/ambari-chef/compare/0.2.0...0.3.0) (Jul 19, 2018)

FEATURES:

- update licenses
- add vdf file upload functionality to ambari_cluster recipe

## [0.2.0](https://github.com/hansohn/ambari-chef/compare/0.1.0...0.2.0) (Jul 18, 2018)

FEATURES:

- bump ambari version to 2.7.0
- bump java version to 1.8.0_181
- add boolean to python recipes to toggle install
- add gpl.license.accepted attribute

BUG FIXES:

- create password.dat file if defining db password

## [0.1.0](https://github.com/hansohn/ambari-chef/compare/0.1.0...0.1.0) (Jun 19, 2018)

FEATURES:

- split out of [hdp-chef](https://github.com/hansohn/hdp-chef)


attributes/default.rb
# -- PYTHON --
default['python'] = {
  'python2' => {
    'install' => true,
    'packages' => ['python'],
    'pips' => [],
  },
}

# -- JAVA --
force_default['java'] = {
  'install_from' => 'amazon_source',
  'install_version' => 'jdk-8u232-linux-x64',
}

# -- AMBARI CLUSTER --
default['hw']['cluster'] = {
  'name' => 'demo',
  'blueprint_name' => 'demo_blueprint',
  'blueprint_file' => 'demo_blueprint.json',
  'hostmapping_file' => 'demo_hostmapping.json',
  'version_definition_file' => 'demo_vdf.json',
}

# -- AMBARI --
default['hw']['ambari'] = {
  'version' => '2.7.3',
  '2.4.0' => {
    'version_full' => '2.4.0.1-1',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.4.0.1/ambari.repo",
  },
  '2.4.3' => {
    'version_full' => '2.4.3.0-30',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.4.3.0/ambari.repo",
  },
  '2.5.0' => {
    'version_full' => '2.5.2.0-298',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.5.2.0/ambari.repo",
  },
  '2.6.2' => {
    'version_full' => '2.6.2.2-1',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.6.2.2/ambari.repo",
  },
  '2.7.0' => {
    'version_full' => '2.7.0.0-897',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.7.0.0/ambari.repo",
  },
  '2.7.3' => {
    'version_full' => '2.7.3.0-139',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.7.3.0/ambari.repo",
  },
  '2.7.4' => {
    'version_full' => '2.7.4.0-118',
    'repo' => "http://public-repo-1.hortonworks.com/ambari/centos#{node['platform_version'].to_i}/2.x/updates/2.7.4.0/ambari.repo",
  },
  'server' => {
    'setup' => {
      'db' => {
        'databasehost' => 'localhost',
        'databaseport' => '5432',
        'databasepassword' => 'bigdata',
      },
      'security' => {
        'setup-truststore' => 'false',
      },
    },
    'config' => {
      'ambari.properties' => {
        'ambari.ldap.isConfigured' => 'false',
        'api.ssl' => 'false',
        'authentication.ldap.bindAnonymously' => 'false',
        'authentication.ldap.dnAttribute' => 'dn',
        'authentication.ldap.groupMembershipAttr' => 'memberUid',
        'authentication.ldap.groupNamingAttr' => 'cn',
        'authentication.ldap.groupObjectClass' => 'posixgroup',
        'authentication.ldap.managerPassword' => '/etc/ambari-server/conf/ldap-password.dat',
        'authentication.ldap.referral' => 'follow',
        'authentication.ldap.usernameAttribute' => 'uid',
        'authentication.ldap.userObjectClass' => 'person',
        'authentication.ldap.useSSL' => 'true',
        'client.api.port' => '8080',
        'client.api.ssl.cert_name' => 'https.crt',
        'client.api.ssl.key_name' => 'https.key',
        'client.api.ssl.port' => '8443',
        'gpl.license.accepted' => 'true',
        'security.server.two_way_ssl' => 'false',
        'server.jdbc.connection-pool' => 'internal',
        'server.jdbc.database_name' => 'ambari',
        'server.jdbc.database' => 'postgres',
        'server.jdbc.driver' => 'org.postgresql.Driver',
        'server.jdbc.hostname' => 'localhost',
        'server.jdbc.port' => 5432,
        'server.jdbc.postgres.schema' => 'ambari',
        'server.jdbc.rca.driver' => 'org.postgresql.Driver',
        'server.jdbc.rca.url' => 'jdbc:postgresql://localhost:5432/ambari',
        'server.jdbc.rca.user.name' => 'ambari',
        'server.jdbc.rca.user.passwd' => '/etc/ambari-server/conf/password.dat',
        'server.jdbc.url' => 'jdbc:postgresql://localhost:5432/ambari',
        'server.jdbc.user.name' => 'ambari',
        'server.jdbc.user.passwd' => '/etc/ambari-server/conf/password.dat',
        'ssl.trustStore.password' => 'changeit',
        'ssl.trustStore.path' => '/var/lib/ambari-server/keys/keystore.jks',
        'ssl.trustStore.type' => 'jks',
      },
    },
    'crypto' => {
      'ca' => '/etc/pki/ca-trust/source/anchors/ca.crt',
      'https_cert' => "/etc/pki/tls/certs/#{node['fqdn']}.crt",
      'https_key' => "/etc/pki/tls/private/#{node['fqdn']}.key",
      'https_key_pass' => 'changeit',
      'https_keystore_p12' => "/etc/pki/tls/certs/#{node['fqdn']}.p12",
      'https_keystore_jks' => "/etc/pki/tls/certs/#{node['fqdn']}.jks",
      'ldap_pass' => 'changeit',
      'truststore_jks' => "/etc/pki/tls/certs/#{node['fqdn']}.jks",
      'truststore_pass' => 'changeit',
    },
    'user' => {
      'name' => 'ambari-server',
      'uid' => '15010',
      'home' => '/var/lib/ambari-server',
      'shell' => '/bin/bash',
    },
  },
  'agent' => {
    'setup' => {
      'security' => {
        'setup-truststore' => 'false',
      },
    },
    'config' => {
      'ambari-agent.ini' => {
        'keysdir' => '/var/lib/ambari-agent/keys',
      },
    },
    'crypto' => {
      'cert' => "/etc/pki/tls/certs/#{node['fqdn']}.crt",
      'key' => "/etc/pki/tls/private/#{node['fqdn']}.key",
      'ca' => '/etc/pki/ca-trust/source/anchors/ca.crt',
      'truststore_jks' => "/etc/pki/tls/certs/#{node['fqdn']}.jks",
    },
    'user' => {
      'name' => 'ambari-agent',
      'home' => '/var/lib/ambari-agent',
      'shell' => '/bin/bash',
      'uid' => '15011',
    },
  },
  'infra' => {
    'config' => {
      'infra-solr-env' => {
        'infra_solr_datadir' => '/opt/ambari_infra_solr/data',
        'infra_solr_keystore_location' => '/etc/security/serverKeys/infra.solr.keyStore.jks',
        'infra_solr_keystore_type' => 'jks',
        'infra_solr_keystore_password' => 'changeit',
        'infra_solr_port' => '8886',
        'infra_solr_ssl_enabled' => 'false',
        'infra_solr_truststore_location' => '/etc/security/serverKeys/infra.solr.trustStore.jks',
        'infra_solr_truststore_type' => 'jks',
        'infra_solr_truststore_password' => 'changeit',
        'infra_solr_user' => 'infra-solr',
      },
    },
    'crypto' => {
      'keystore_jks' => '/var/lib/ambari-infra-solr/keys/infra.solr.keyStore.jks',
      'truststore_jks' => '/var/lib/ambari-infra-solr/keys/infra.solr.trustStore.jks',
    },
    'user' => {
      'name' => 'infra-solr',
      'home' => '/home/infra-solr',
      'shell' => '/bin/bash',
      'uid' => '15018',
    },
  },
  'metrics' => {
    'config' => {
      'ams-grafana-env' => {
        'metrics_grafana_data_dir' => '/var/lib/ambari-metrics-grafana',
        'metrics_grafana_username' => 'admin',
        'metrics_grafana_password' => 'admin',
      },
      'ams-grafana-ini' => {
        'cert_file' => '/etc/ambari-metrics-grafana/conf/ams-grafana.crt',
        'cert_key' => '/etc/ambari-metrics-grafana/conf/ams-grafana.key',
        'port' => '3000',
        'protocol' => 'http',
      },
      'ams-hbase-site' => {
        'hbase.rootdir' => 'file:///var/lib/ambari-metrics-collector/hbase',
        'hbase.tmp.dir' => '/var/lib/ambari-metrics-collector/hbase-tmp',
      },
      'ams-site' => {
        'timeline.metrics.aggregator.checkpoint.dir' => '/var/lib/ambari-metrics-collector/checkpoint',
      },
    },
    'crypto' => {
      'cert' => '/etc/ambari-metrics-grafana/conf/ams-grafana.crt',
      'key' => '/etc/ambari-metrics-grafana/conf/ams-grafana.key',
    },
    'user' => {
      'name' => 'ams',
      'home' => '/home/ams',
      'shell' => '/bin/bash',
      'uid' => '15013',
    },
    'group' => {
      'name' => 'hadoop',
      'gid' => '10010',
    },
  },
}

# -- HADOOP --
default['hw']['hadoop'] = {
  'common' => {
    'group' => {
      'name' => 'hadoop',
      'gid' => '10010',
    },
  },
}

# -- LOGSEARCH --
default['hw']['logsearch'] = {
  'config' => {
    'logfeeder-env' => {
      'logfeeder_keystore_location' => '/etc/security/serverKeys/logsearch.keyStore.jks',
      'logfeeder_keystore_type' => 'jks',
      'logfeeder_keystore_password' => 'changeit',
      'logfeeder_truststore_location' => '/etc/security/serverKeys/logsearch.trustStore.jks',
      'logfeeder_truststore_type' => 'jks',
      'logfeeder_truststore_password' => 'changeit',
    },
    'logsearch-admin-json' => {
      'logsearch_admin_username' => 'ambari_logsearch_admin',
      'logsearch_admin_password' => 'ambari_logsearch_password',
    },
    'logsearch-env' => {
      'logsearch_keystore_location' => '/etc/security/serverKeys/logsearch.keyStore.jks',
      'logsearch_keystore_type' => 'jks',
      'logsearch_keystore_password' => 'changeit',
      'logsearch_truststore_location' => '/etc/security/serverKeys/logsearch.trustStore.jks',
      'logsearch_truststore_type' => 'jks',
      'logsearch_truststore_password' => 'changeit',
      'logsearch_ui_port' => '61888',
      'logsearch_ui_protocol' => 'http',
    },
  },
  'crypto' => {
    'keystore_jks' => '/var/lib/logsearch/keys/logsearch.keyStore.jks',
    'truststore_jks' => '/var/lib/logsearch/keys/logsearch.trustStore.jks',
  },
  'user' => {
    'name' => 'logsearch',
    'home' => '/home/logsearch',
    'shell' => '/bin/bash',
    'uid' => '15021',
  },
}

# -- AMBARI LDAP --
if defined?(node.chef_environment)
  case node.chef_environment
  when 'production'
    override['hw']['ambari']['server']['config']['ambari.properties']['host_cname'] = 'ambari.prd.domain.local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.baseDn'] = 'dc=prd,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.managerDn'] = 'uid=manager,cn=users,cn=accounts,dc=prd,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.primaryUrl'] = 'ldap-server-01.prd.domain.local:636'
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_groups'] = ['prd_user_group']
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_users'] = ['prd_user']
  when 'staging'
    override['hw']['ambari']['server']['config']['ambari.properties']['host_cname'] = 'ambari.stg.domain.local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.baseDn'] = 'dc=stg,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.managerDn'] = 'uid=manager,cn=users,cn=accounts,dc=stg,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.primaryUrl'] = 'ldap-server-01.stg.domain.local:636'
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_groups'] = ['stg_user_group']
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_users'] = ['stg_user']
  else
    override['hw']['ambari']['server']['config']['ambari.properties']['host_cname'] = 'ambari.dev.domain.local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.baseDn'] = 'dc=dev,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.managerDn'] = 'uid=manager,cn=users,cn=accounts,dc=dev,dc=domain,dc=local'
    override['hw']['ambari']['server']['config']['ambari.properties']['authentication.ldap.primaryUrl'] = 'ldap-server-01.dev.domain.local:636'
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_groups'] = ['dev_user_group']
    override['hw']['ambari']['server']['setup']['ldap']['ldap_sync_users'] = ['dev_user']
  end
end
